### DISCUSSION TITLE
update CCRotaryEncoder enc multiplier with a button press

### DISCUSSION THREAD

**Discussion by IQworkshop** [tttapa/Control-Surface#754] [Q&A] [answered]
Hi,

Newbie programmer here, so I'll apologise for the dumb question ahead of time.
I am attempting to build a control surface to interface with premiere pro (practical application for learning).

Currently, I have the scrub wheel and 3 buttons working (REW,play/pause/FF)
what I would like to be able to do is update the scrub wheel's multiplier with a button press, so I can have a variable resolution when scrubbing media.

I see from the documentation there is a function (setspeedMultiply()), but I cannot get this to compile without errors.
Any suggestions would be appreciated.

main.cpp
```cpp
#include <Arduino.h>
#include <Encoder.h>
#include <Control_Surface.h>  // Include the Control Surface library

 
// Instantiate a MIDI over USB interface.
USBMIDI_Interface midi;

int8_t rotarySensitivity = 8;
// Instantiate a CCRotaryEncoder object
CCRotaryEncoder enc {
  {2, 3},       // pins
  MIDI_CC::Scrub_Wheel, // MIDI address (CC number + optional channel)
  rotarySensitivity,            // optional multiplier if the control isn't fast enough
};


// define buttons
Button pushbutton1 {9};
Button pushbutton2 {8};
Button pushbutton3 {7};
Button pushbutton4 {6};

#include <Functions.h>        //include functions lib

void setup() {
  // Set the relative MIDI CC mode.
  RelativeCCSender::setMode(relativeCCmode::MACKIE_CONTROL_RELATIVE);
  Control_Surface.begin(); // Initialize Control Surface

  }
 
void loop() {
   procButton();
   Control_Surface.loop(); // Update the Control Surface
}
```

Functions.h
```cpp
void procButton() {
    
    const uint8_t velocity = 127; 
    if (pushbutton1.update() == Button::Falling) {midi.sendNoteOn(0x5b, velocity);}
    if (pushbutton2.update() == Button::Falling) {  /*trigger change of resolution using switch case*/      }
    if (pushbutton3.update() == Button::Falling) {midi.sendNoteOn(0x5e, velocity);}
    if (pushbutton4.update() == Button::Falling) {midi.sendNoteOn(0x5c, velocity);}
}
```

**Comment by tttapa** [comment]
> I see from the documentation there is a function (setspeedMultiply()), but I cannot get this to compile without errors.
> Any suggestions would be appreciated.

I don't see it used in the code you posted? What did you try and what error message did you get?

As a side note: it's usually not a good idea to define (non-inline) functions in a header (.h) file. Why is `procButton()` in a separate file?

**Comment by IQworkshop** [comment, marked as answer]
I think I have found a solution to my challenge.
The error I was getting was PlatformIO was not recognising the function call, also I do recall seeing one error stating the function I was calling at that time was private. I've since deleted that code and had abandoned the idea pending the output of this question, so I can't recall the details clearly.

I have used the code at the bottom of this post [https://forum.arduino.cc/t/relative-rotary-encoders-control_surface-h-help-needed/663200/8](url)

I made the following change to my procButton() call

`
if (pushbutton2.update() == Button::Falling) {  
        switch (rotarySensitivity)
        {
        case 1:
            rotarySensitivity = 8;
            break;
        
        case 8:
            rotarySensitivity = 16;
            break;
        
        case 16:
            rotarySensitivity = 32;
            break;
        
        case 32:
            rotarySensitivity = 64;
            break;
        
        case 64:
            rotarySensitivity = 1;
            break;
        default:
            break;
        }

        enc.setMultiplier(rotarySensitivity) ;     
    }
`
This seems to be working now

With regards to the procButton() function in a separate file. That is just my style of writing that makes sense to me. I struggle to keep track of, and make sense of code when it's in a single file. something about scrolling up and down the code form the definition to the call really spins my thought process out.

Breaking my functions into a separate file means I know exactly where they are, and I can tab between pages and keep my train of thought. It may not make sense to some, and not propper programming etiquette, but makes sense to me.

I'm not a professional programmer (clearly), but I'm likely to be the only one reading the code unless im reaching out for help.
I swear I'm undiagnosed ADHD lol

Thanks for this awesome project!
